File: backend/app/schemas/notifications.py
"""Strict source contract and authenticated notification API contracts."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator, model_validator

from app.schemas.finding import Pagination

Locale = Literal["de", "da", "en"]
NotificationType = Literal["unpublished_upcoming_event", "quality_finding"]
NotificationStatus = Literal["pending", "active", "resolved", "suppressed", "expired"]
DeliveryKind = Literal["initial", "reminder", "escalation", "digest", "test"]
DeliveryStatus = Literal["queued", "sending", "sent", "failed", "permanent_failure", "cancelled"]


class StrictConfig(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True, hide_input_in_errors=True)


class NotificationRecipient(StrictConfig):
    email: EmailStr
    locale: Locale = "de"
    enabled: bool = True


class ReminderConfig(StrictConfig):
    enabled: bool = False
    days_after: int = Field(default=7, ge=1, le=90)


class UnpublishedConfig(StrictConfig):
    enabled: bool = False
    days_before: int = Field(default=14, ge=1, le=365)
    reminder: ReminderConfig = Field(default_factory=ReminderConfig)


class QualityConfig(StrictConfig):
    enabled: bool = False
    minimum_severity: Literal["info", "warning", "error"] = "warning"
    delivery: Literal["daily_digest"] = "daily_digest"


class NotificationEvents(StrictConfig):
    unpublished_upcoming_events: UnpublishedConfig = Field(default_factory=UnpublishedConfig)
    quality_findings: QualityConfig = Field(default_factory=QualityConfig)


class NotificationConfig(StrictConfig):
    version: Literal[1]
    recipients: list[NotificationRecipient] = Field(max_length=100)
    events: NotificationEvents

    @field_validator("version", mode="before")
    @classmethod
    def integer_version(cls, value: object) -> object:
        if type(value) is not int:
            raise ValueError("Version must be an integer")
        return value

    @model_validator(mode="after")
    def unique_recipients(self) -> "NotificationConfig":
        addresses = [str(item.email).casefold() for item in self.recipients]
        if len(addresses) != len(set(addresses)):
            raise ValueError("Duplicate recipient")
        return self


class NotificationPayload(BaseModel):
    """Small, typed rendering/audit snapshot; never a raw source row or finding.message."""

    organization_name: str
    entity_name: str
    entity_type: str
    entity_key: str
    action_path: str
    event_status: Literal["draft", "review"] | None = None
    next_date: str | None = None
    days_until: int | None = None
    stage: int = 0
    episode: int = Field(default=1, ge=1)
    rule: str | None = None
    finding_id: str | None = None
    severity: str | None = None
    field: str | None = None
    relevance: list[str] = Field(default_factory=list)
    priority: Literal["urgent", "important", "improvement", "internal"] = "improvement"


class NotificationSummary(BaseModel):
    id: UUID
    notification_type: NotificationType
    organization_id: UUID
    entity_type: str | None
    entity_key: str | None
    entity_name: str | None
    finding_id: str | None
    rule: str | None
    status: NotificationStatus
    payload: NotificationPayload
    first_detected_at: datetime
    last_detected_at: datetime
    resolved_at: datetime | None
    expired_at: datetime | None


class NotificationDelivery(BaseModel):
    id: UUID
    organization_id: UUID
    recipient: str
    locale: Locale
    delivery_kind: DeliveryKind
    status: DeliveryStatus
    subject: str | None
    message_fingerprint: str
    attempt_count: int
    queued_at: datetime | None
    sending_at: datetime | None
    sent_at: datetime | None
    next_attempt_at: datetime | None
    last_error: str | None
    provider_message_id: str | None


class NotificationDetail(NotificationSummary):
    deliveries: list[NotificationDelivery]
    delivery_enabled: bool


class DeliveryDetail(NotificationDelivery):
    notifications: list[NotificationSummary]


class ConfigIssue(BaseModel):
    organization_id: UUID
    organization_name: str
    code: Literal["invalid_config"] = "invalid_config"


class NotificationHealth(BaseModel):
    delivery_enabled: bool
    source_capability: bool
    config_issues: list[ConfigIssue]


class NotificationCounts(BaseModel):
    active: int
    queued: int
    sent_today: int
    failed: int


class NotificationPage(BaseModel):
    items: list[NotificationSummary]
    pagination: Pagination
    summary: NotificationCounts
    health: NotificationHealth


class NotificationPreview(BaseModel):
    subject: str
    text: str
    html: str
    locale: Locale
    notification_ids: list[UUID]
    delivery_enabled: bool

File: frontend/tests/fixtures/notifications.ts
import type {
  NotificationSummary,
  NotificationDelivery,
  NotificationPage,
} from '../../shared/contracts'
export const notification: NotificationSummary = {
  id: '10000000-0000-4000-8000-000000000031',
  notification_type: 'unpublished_upcoming_event',
  organization_id: '10000000-0000-4000-8000-000000000010',
  entity_type: 'event',
  entity_key: '10000000-0000-4000-8000-000000000021',
  entity_name: 'Kulturabend',
  finding_id: null,
  rule: null,
  status: 'active',
  payload: {
    organization_name: 'Kulturverein',
    entity_name: 'Kulturabend',
    entity_type: 'event',
    entity_key: '10000000-0000-4000-8000-000000000021',
    internal_action_path: '/events/10000000-0000-4000-8000-000000000021',
    external_action_url:
      'https://app.kulturbytes.de/admin/event/10000000-0000-4000-8000-000000000021',
    event_status: 'draft',
    next_date: '2026-09-30',
    days_until: 12,
    stage: 1,
    episode: 1,
    rule: null,
    finding_id: null,
    severity: null,
    field: null,
    relevance: [],
    priority: 'improvement',
  },
  first_detected_at: '2026-09-18T10:00:00Z',
  last_detected_at: '2026-09-18T10:00:00Z',
  resolved_at: null,
  expired_at: null,
}
export const notificationDelivery: NotificationDelivery = {
  id: '10000000-0000-4000-8000-000000000041',
  organization_id: notification.organization_id,
  recipient: 'recipient@example.test',
  locale: 'da',
  delivery_kind: 'initial',
  status: 'failed',
  subject: 'Dit arrangement finder snart sted',
  message_fingerprint: 'a'.repeat(64),
  attempt_count: 1,
  queued_at: notification.first_detected_at,
  sending_at: notification.first_detected_at,
  sent_at: null,
  next_attempt_at: '2026-09-18T10:05:00Z',
  last_error: 'smtp_451',
  provider_message_id: null,
}
export const notificationPage: NotificationPage = {
  items: [notification],
  pagination: { page: 1, page_size: 50, total: 1, pages: 1 },
  summary: { active: 1, queued: 0, sent_today: 0, failed: 1 },
  health: { delivery_enabled: false, source_capability: true, config_issues: [] },
}
export const notificationDetail = {
  ...notification,
  deliveries: [notificationDelivery],
  delivery_enabled: false,
}
export const notificationDeliveryDetail = { ...notificationDelivery, notifications: [notification] }
export const notificationGuidance = {
  de: 'Melde dich bei Kulturbytes an und öffne den entsprechenden Eintrag in deinem Dashboard.',
  da: 'Log ind på Kulturbytes, og åbn det relevante opslag i dit dashboard.',
  en: 'Sign in to Kulturbytes and open the relevant entry in your dashboard.',
}
export function notificationPreview(locale: 'de' | 'da' | 'en', withAction = true) {
  const subject = {
    de: 'Deine Veranstaltung „Kulturabend“ findet bald statt',
    da: 'Dit arrangement „Kulturabend“ finder snart sted',
    en: 'Your event “Kulturabend” is coming up soon',
  }[locale]
  const url = notification.payload.external_action_url
  const label = { de: 'Veranstaltung bearbeiten', da: 'Rediger arrangementet', en: 'Edit event' }[
    locale
  ]
  const action = withAction ? `<a href="${url}">${label}</a>` : notificationGuidance[locale]
  return {
    subject,
    text: `${subject}\nKulturverein\n${withAction ? `${label}: ${url}` : notificationGuidance[locale]}`,
    html: `<!doctype html><html lang="${locale}"><head></head><body><p>${subject}</p><p>${action}</p></body></html>`,
    locale,
    notification_ids: [notification.id],
    delivery_enabled: false,
  }
}
